Untyped_result_set: make non-copying and fragment retaining

elcallio opened the pull request:
Refs #7961
Fixes #8014

The "untyped_result_set" object was created for small, internal access to cql-stored metadata.
It is nowadays used for rather more than that (cdc). 
This has the potential of mixing badly with the fact that the type does deep copying of data
and linearizes all (not to mention handles multiple rows rather inefficiently).

Instead of doing a deep copy of input, we keep assume ownership and build
rows of the views therein, potentially retaining fragmented data as-is
avoiding premature linearization.

Note that this is not all sugar and flowers though. Any data access will
by nature be more expensive, and the view collections we create are
potentially just as expensive as copying for small cells.

Otoh, it allows writing code using this that avoids data copying,
depending on destination.

v2:
* Fixed wrong collection reserved in visitor
* Changed row index from shared ptr to ref
* Moved typedef
* Removed non-existing constructors
* Added const ref to index build
* Fixed raft usage after rebase

v3:
* Changed shared_ptr to unique

haaawk commented on cql3/untyped_result_set.hh:
```diff
@@ -41,36 +41,61 @@
+    friend class untyped_result_set;
+    using index_map = std::unordered_map<std::string_view, size_t>;
+    using data_views = std::vector<view_holder>;
+    using index_map_ptr = ::shared_ptr<index_map>;
```
could this be a lw_shared_ptr instead of ::shared_ptr?

elcallio replied:
I guess. The types are known, so... Could even be a ref + unique_ptr in parent object.

haaawk commented on cql3/untyped_result_set.cc:
```diff
@@ -47,58 +47,118 @@
+        [](std::monostate) -> view_type { throw std::bad_variant_access(); },
+        [](const view_type& v) -> view_type { return v; },
+        [](const bytes& b) -> view_type { return view_type(b); }
+    ), _data.at(index(name)));
```
I never know whether we should be using at() and pay for safety or [] and avoid checks. @avikivity what's your take on this?

elcallio replied:
When things are not in hot path, I prefer the checked accessors, just to make sure we catch any borked code.

avikivity replied:
Agree with @elcallio. These less-hot code paths are also more error-prone.

haaawk replied:
Isn't this code executed at least once for each request? I thought that whatever gets executed 'per request' is already a hot path, no? I understand that we have a more hot path when something gets executed 'per sstable' but still once per request can be 1-5M times / s, no?

elcallio replied:
That is not a hot path. An inner loop is a hot path.

haaawk replied:
It seems that every person has its own definition of a hot path :). @tgrabiec was always telling me 'per request' operation is a hot path and I sort of agree considering that we're aiming for 5M ops/s

elcallio replied:
That depends on the context. For a bounds check it is not. For a major allocation or prime number calculations it might be.

haaawk commented on cql3/untyped_result_set.cc:
```diff
@@ -47,58 +47,118 @@
+    auto map = ::make_shared<untyped_result_set_row::index_map>();
+    auto& names = metadata.get_names();
+    size_t i = 0;
+    std::transform(names.begin(), names.end(), std::inserter(*map, map->end()), [&](lw_shared_ptr<column_specification> c) mutable {
```
nit: I think const lw_shared_ptr<column_specification>& would work as well.

haaawk commented on cql3/untyped_result_set.hh:
```diff
@@ -188,8 +217,19 @@ public:
+        , ::shared_ptr<cql_transport::messages::result_message>
+        , qr_tuple
+    >;
+    struct visitor;
```
Why do we have to expose this visitor in the header? Wouldn't it be enough to make it internal to .cc file?

elcallio replied:
namespacing + private types + friend relationship.

haaawk replied:
I don't follow. Could you elaborate a bit more for weaker audience ;).

elcallio replied:
The visitor/static below use types and constructors that I made private in the row type (because they should be internal). untyped_result_set can see them because it is a friend. Top-level routines/types cannot. Can make more elaborate friend declarations, but to me at least it seems easier to just do these harmless (and private) declarations instead. 
Also, I prefer nested types to file-local ones, if for nothing else so for scoping of concern.

haaawk replied:
Got you. I was looking at a wrong friendship relation (result_generator <-> untyped_result_set) and couldn't figure out why it matters :)

avikivity replied:
I agree about nesting, but the #include dependency addition is unfortunate. Maybe hide the variant behind a unique_ptr? It's just a placeholder after all, no?

Also, std::any? probably doesn't work due to copyablity.

elcallio replied:
I felt that yet another level of indirection was perhaps pushing it... Not happy about the extra indirection - really with the foreign_ptr was not so an lw_shared_ptr... Even tried to work around the forward declaration aspect there, but it just becomes terrible. 
std::any does not work because of move-only types.

haaawk commented on cql3/untyped_result_set.hh:
```diff
@@ -188,8 +217,19 @@ public:
     rows_type _rows;
+
     untyped_result_set() = default;
+    static untyped_result_set_row::index_map_ptr make_index(const cql3::metadata&);
```
I know it is a matter of taste but maybe it would be better to hide this function inside .cc file and not expose it in the header since it's only used internally?

elcallio replied:
same as above.

haaawk commented on cql3/untyped_result_set.hh:
```diff
@@ -147,23 +172,27 @@ public:
     untyped_result_set(::shared_ptr<cql_transport::messages::result_message>);
-    untyped_result_set(const cql3::result_set&);
+    untyped_result_set(const schema&, foreign_ptr<lw_shared_ptr<query::result>>, const cql3::selection::selection&, const query::partition_slice&);
+    untyped_result_set(cql3::result_set&&);
+    untyped_result_set(cql3::result&&);
```
Where are those two constructors implemented? I couldn't find their definitions. Are they used at all?

elcallio replied:
They are not supposed to be there. Sorry, forgot to prune...

haaawk commented on cql3/untyped_result_set.cc:
```diff
@@ -47,58 +47,118 @@
+    {}
+
+    void start_row() {
+        rows.reserve(index->size());
```
Shouldn't this be tmp.reserve(...)?

elcallio replied:
Very much so. Me so sloppy.

haaawk commented on cql3/untyped_result_set.cc:
```diff
@@ -47,58 +47,118 @@
+            tmp.emplace_back(std::monostate{});
+        } 
+    }
+    // somewhat weird dispatch, but when visiting directly via
+    // result_generator, pk:s will be temporary - and sent 
+    // as views, not opt_views. So we can dispatch on this and 
+    // simply copy the temporaries.
```
Would it be possible to move from those temporaries instead of copying from them?

elcallio replied:
Maybe. Not sure if that violates the contract.

elcallio replied:
No, you cannot, since the args are views, not the actual temporaries. And changing that in turn would break other code, which I guess is fixable, but probably not worth it... At least not for now.

haaawk reviewed (approved):
LGTM

avikivity commented on types/listlike_partial_deserializing_iterator.hh:
```diff
@@ -28,6 +28,12 @@
+
+template <FragmentedView View>
+View read_collection_value(View& in, cql_serialization_format);
+
```
How does it work? These are templates, if they are defined in a source file, then linkage may break if used with the wrong types. If they are defined in a header file, then this is unneeded.

Is it just needed as a forward declaration?

elcallio replied:
These templates are explicitly instantiated in types.cc, so magically enough it works (no compile/linkage errors, even without an extern template decl - same as collection type deserialize templates).

avikivity replied:
ok.

elcallio commented:
Updated and rebased
v2:
* Fixed wrong collection reserved in visitor
* Changed row index from shared ptr to ref
* Moved typedef
* Removed non-existing constructors
* Added const ref to index build
* Fixed raft usage after rebase

kbr- commented on cql3/untyped_result_set.hh:
```diff
@@ -188,8 +214,20 @@ public:
+    struct visitor;
+
+    storage _storage;
+    index_map_ptr _index;
```
Does this have to be a pointer? If I understand correctly, both `untyped_result_set::visitor` and `untyped_result_set_row` lifetimes are contained in `untyped_result_set`'s lifetime, so `untyped_result_set` could just store an `index_map` directly and these guys would have a reference to it.

elcallio replied:
I want both types to be movable. Having the index stored inside untyped_result_set wound break refs in the row objects.

kbr- replied:
Then `unique_ptr`?

elcallio replied:
I guess it can be. now that we made the refs in rows actual refs. Not a huge deal though.

elcallio replied:
changed.

kbr- reviewed (approved):
one nit/question, otherwise LGTM

kbr- commented on cql3/untyped_result_set.cc:
```diff
@@ -47,58 +47,118 @@
+struct cql3::untyped_result_set::visitor {
+    rows_type& rows;
+    const cql3::metadata& meta;
+    const index_map_ptr& index;
```
Now moving `untyped_result_set` will cause a dangling reference, no?
I had something different in mind, i.e. `const index_map&`, which is a reference to whatever the `unique_ptr` points to (and that something doesn't move when we move the owning `untyped_result_set`)

kbr- replied:
Ok nvm it's just the `visitor`, `untyped_result_set` doesn't move during the lifetime of the `visitor` so there's no problem.
But still, why `ptr&` instead of just `&`

elcallio replied:
Smallest possible change. And avoid using long type reference.

kbr- replied:
k

elcallio commented:
ping?

elcallio commented:
ping?

slivne commented:
@avikivity kamil and piotrj blessed this - and calle addressed your comments - ping

avikivity commented:
Dequeued, unit tests don't compile.

elcallio commented:
Rebased. Fixed implicit linkage error.
